### DISCUSSION TITLE
Javascript functions security

### DISCUSSION THREAD

**Discussion by zane-commeatio** [vercel/next.js#46178] [Help]
Hi,, i wonder if someone could call a function programmatically.

Example:
I create a function `generateBalanceSheet()` in `/utils/admin/critical.ts`
This function is imported only in an api route `/api/admin/gen-bs.ts` which is gated by passing a token in the request headers.

Would this function be bundled in the client side app ? an callable by some shadow way ?

**Comment by icyJoseph** [comment]
Well, even if we say it is possible, they'd still need the token, and probably other Node built-ins that are not available client side, we could speculate that anyway.

If the function is only consumed by API routes, then it is good.

Next.js runs three builds, client, server and edge. You can see this by creating a webpack key on next.config.js and see the `nextRuntime`:

```js
  webpack(config, { nextRuntime }) {
    console.log(nextRuntime); // undefined, nodejs, edge
    return config;
  },
```

Tree shaking will also kick in and remove unused code. The only way I can think of to leak this code is through `dynamic imports`, where one basically makes up a bundle out of the contents of a file, and then it could be send to the client, but that's a really long stretch.

This tool https://next-code-elimination.vercel.app/ can help you with pages as well.

That being said, you'd need to explicitly in some way include the code, it won't just happen by mistake. 

```js
// utils/secrets.ts
export const foo = () => {
  console.log("I should not leak");
};

export const bar = () => {
  console.log("I am safe");
};
```

The following might cause leaks, for example if you uncomment:  `const leak = () => import("../utils/secrets");`, and let it hang, it'll leak, but if you use it within `getStaticProps`, it'll be removed. Try various combinations.

```js
import Head from "next/head";
import { useEffect } from "react";
// import { foo } from "../utils/secrets";
import { bar } from "../utils/secrets";
// const leak = () => import("../utils/secrets");

export default function Home() {
  // useEffect(() => {
  //   leak();
  // }, []);

  return (
    <div>
      <Head>
        <title>Home title</title>
      </Head>

      <h1>home</h1>
    </div>
  );
}

export const getStaticProps = () => {
  // leak();
  return { props: {} };
};

```

However, the bottom line is that it won't leak just by accident.

**Comment by zane-commeatio** [comment]
Thanks you that runtime check is very much what i needed 🙏
